task_processing/plugins/mesos/logging_executor.py

```diff
@@ -0,0 +1,151 @@
+        # Do not fetch logs if url hasn't been built
+        if log_md.log_url is '':
+            return
+
+        offsets = {
+            'stdout': log_md.log_offsets['stdout'],
+            'stderr': log_md.log_offsets['stderr']
+        }
+        for f in ['stdout', 'stderr']:
+            offset = offsets[f]
+            while True:
+                url = log_md.log_url + f + \
+                    '&length=' + str(TASK_LOG_CHUNK_LEN) + \
+                    '&offset=' + str(offset)
+                try:
+                    response = json.load(urllib.request.urlopen(url))
+                    log_length = len(response['data'])
+                    for line in response['data'].splitlines():
+                        print(
+                            task_id + ": " + line,
+                            file=sys.stderr if f is 'stderr' else sys.stdout
+                        )
+                    offset = offset + log_length
+                    # Stop if there is no more data
+                    if log_length < TASK_LOG_CHUNK_LEN:
+                        break
+                except Exception as e:
+                    log.error("Failed to fetch task {f} log {error}".format(
+                        f=f, error=e
+                    ))
+            # Update offset of this stream
+            offsets[f] = offset
+
+        # Update both offsets for the task
+        self.log_metadata = self.log_metadata.set(
+            task_id,
+            log_md.set(log_offsets=pmap(offsets)),
+        )
+
+    def logging_loop(self):
+        while True:
+            # process downstream events
+            while not self.src_queue.empty():
+                e = self.src_queue.get()
+                self.dest_queue.put(e)
+                # Record the 1st part of the log url
+                if e.kind == 'task' and e.platform_type == 'staging':
+                    url = e.raw.scheme + '://' + \
+                        e.raw.ip + ':' + \
+                        str(e.raw.port) + '/'
+                    self.log_metadata = self.log_metadata.set(
+                        e.task_id,
+                        LogMetadata(log_url=url)
+                    )
+
+                # Build the rest of the log url
+                if e.kind == 'task' and e.platform_type == 'running':
+                    log_md = self.log_metadata[e.task_id]
+                    url = self.build_task_log_url(log_md.log_url, e)
+                    self.log_metadata = self.log_metadata.set(
+                        e.task_id,
+                        log_md.set(log_url=url)
+                    )
+
+                # Fetch the last log and remove the entry
+                if e.kind == 'task' and e.terminal:
+                    self.stream_task_log(e.task_id)
+                    self.log_metadata.discard(e.task_id)
+
+            # grab logs
+            for task_id in self.log_metadata.keys():
+                self.stream_task_log(task_id)
```

Comment:
i don't think we can afford to fetch logs in this thread as we'd be blocking the queue chain

Reply:
Good catch! I will fix it along with other comments from you and @Rob-Johnson.